### ISSUE TITLE
add gallium nine standalone for option

### ISSUE THREAD

**Issue by Meklahi** [lutris/lutris#1625]
[https://github.com/dhewg/nine](url)

Gallium Nine Standalone, as the name implies, is a standalone version of the WINE parts of Gallium Nine.
This decouples Gallium Nine from the WINE tree, so that it can be used with any WINE version. There is no need for any WINE patches. A stable, development, or staging WINE release is sufficient.
WINE can be updated independent of Gallium Nine Standalone. Users can mix releases of both projects to their liking. Switching between staging and non-staging does not require a rebuild.

**Comment by romulasry** [comment]
This would be nice to have. A toggle button would be cool.

**Comment by strycore** [comment]
the problem is getting a gallium9 build, once that is sorted out, it's easy to add the option in lutris.

**Comment by dhewg** [comment]
Oh nice, just stumbled upon this issue!

The project moved: https://github.com/iXit/wine-nine-standalone
Since v0.2 we also provide binary releases: https://github.com/iXit/wine-nine-standalone/releases

Those can be installed by the bundled script, see https://github.com/iXit/wine-nine-standalone#usage
Or you can do it by hand, like winetricks: https://github.com/Winetricks/winetricks/commit/2a4ec13d86132fbe1c69afa75bd0981f45f15af0

Does that work for you guys?

**Comment by dhewg** [comment]
Is there anything that needs to be improved for the standalone version to get this into lutris?

**Comment by sonicpp** [comment]
@strycore Do you want to provide custom build or rely on gallium-nine provided by distribution package (like Lutris already does with Feral gamemode and FPS limiter for example)?

I can start working on this if you are busy with other features...

**Comment by strycore** [comment]
@sonicpp yeah, I have very little experience with Gallium nine so that would be very helpful. We build wine with this script https://github.com/lutris/buildbot/blob/master/runners/wine/build.sh but this requires a specific setup (it's explained in the readme of the buildbot repo)

I'll push that to 0.5.3 to give more time but if we get progress made earlier this might be ready for 0.5.2 (or maybe there's no work to be done on the client side at all and a custom build will be enough)

**Comment by sonicpp** [comment]
Hello,

So I have finally time to work on this.

At the moment I rely on gallium-nine provided by distribution, but later we can provide custom build as part of Lutris.

So far I investigated following (I already coded simple implementation and tested it on my localhost for 32 bit prefix):
* installing Nine to current prefix - copy _d3d9-nine.dll_ and _ninewinecfg.exe_ provided by distribution package into C:\Windows\System32 of current (32 bit) prefix
* enabling Nine - To enable Gallium Nine one has to run command `ninewinecfg.exe -e` or enable it manually (we will lost the original d3d9.dll file in system32, but it probably does not matter):
  * set key: HKEY_CURRENT_USER : Software\\Wine\\DllOverrides : d3d9 to _native_
  * create symlink C:\Windows\System32\d3d9.dll -> C:\Windows\System32\d3d9-nine.dll
* disabling Nine - To disable Nine, simply run `ninewinecfg.exe -d` or do following:
  * delete key: HKEY_CURRENT_USER : Software\\Wine\\DllOverrides : d3d9
  * delete file C:\Windows\System32\d3d9.dll
* checking current status: there is no util for this, but we can do it manually. Gallium Nine is enabled in following conditions:
  * HKEY_CURRENT_USER : Software\\Wine\\DllOverrides : d3d9 is _native_
  * C:\Windows\System32\d3d9.dll is symlink **and** file C:\Windows\System32\d3d9.dll exists

So far it is simple task, but one also need to know if Gallium Nine is supported on current system. Gallium Nine requires following setup:
* MESA with Gallium Nine support (checking /usr/lib/d3d/d3dadapter9.so or similar path should be enough)
* Installed Gallium Nine Standalone (may be installed with Lutris in future), one may check if _d3d9-nine.dll_ and _ninewinecfg.exe_ are present in system directory (/usr/lib32/wine/ for example)
* Chosen MESA driver supports Gallium Nine - ninewinecfg is checking this by calling _Direct3DCreate9Ex()_ function from _d3d9-nine.dll_. I think we will need to do the same.

TLDR - the biggest problem I have at the moment is checking current MESA driver against Gallium Nine. The problem is that user can enable Gallium Nine when he has selected AMD GPU (by DRI_PRIME) and then start game with different graphics card (like disabling DRI_PRIME in Lutris, which will activate Intel GPU for my setup for example). The game will not be able to create D3D9 device - in this situation Lutris should fall back to setup with disabled Gallium Nine.

How to handle this situation? Check result of _Direct3DCreate9Ex_ every time the game starts (we will need simple WIN32 binary for that) with enabled Gallium Nine, and disable it in case of error? Or do a list of available graphics cards on user's PC (Lutris is already doing in startup.check_driver) and test every GPU for Gallium Nine support on startup (that means calling _Direct3DCreate9Ex_ during startup only and not every time some game is started with enabled Gallium Nine support)?

@dhewg any suggestions?

**Comment by dhewg** [comment]
IMHO it'd be best if lutris would use `ninewinecfg` with the current `-e`/`-d` switches.

Then we could introduce a new probe switch like `-p`, which tests if the current setup is capable of using nine. That switch can be implemented so it works independent of the current activation status.

Lutris can then use `-p` to check the setup (if it fails maybe even try again with `DRI_PRIME`), and only allow to enable nine if that passes.

With that, getting nine into a WINE prefix is left. But that's just a matter of symlinking or copying files.

**Comment by sonicpp** [comment]
Cool, my PR already uses `ninewinecfg` with `-e` and `-d` switches. Once `-p` is available, I will update my PR (or create a new one, if it gets merged before).

**Comment by dhewg** [comment]
@sonicpp Is the `-p` switch still of interest?

**Comment by dhewg** [comment]
@sonicpp 
The just released v0.5 includes said `-p` switch. The nine binaries need to be part of a WINE prefix, but nine doesn't have to be enabled to use it.

The exit code is `0` if the current setup is nine-capable. If I use it on my iris laptop it correctly returns `1` as the mesa iris driver is not yet the default (and the classic driver is not a gallium driver). If I use it with `MESA_LOADER_DRIVER_OVERRIDE=iris` it return `0` though ;)

The same should work with `DRI_PRIME=1` on such setups. You decide if lutris should probe with that if the initial probe fails ;)

**Comment by sonicpp** [comment]
@dhewg thanks! I will test it as soon as possible (nine support was merged to lutris, but there is one issue I have to fix - it is not issue of nine itself)

**Comment by SeerLite** [comment]
Hey, just wanted to bump this and ask if Gallium Nine is officially supported? I'm asking because this issue is still open, yet the option is available from within Lutris?
